### ISSUE TITLE
Explicit param passing syntax

### ISSUE THREAD

**Issue by oliverbrooks** [orangemug/sql-stamp#9]
What do you think of passing params to sql requires?

Aim would be to explicitly tell the requires what variables they will receive and avoid mixed use of variables. This becomes especially useful when there are lots of nested requires or you want to use something generic like `id` or `userId` for different things across files.

Seems like it would be more 'nodey' self contained file type thing.

```js
var sqlStamp = require("sql-stamp");

sql("foo.sql", {
  id: 5, 
  badger: "badger"
})
```

foo.sql

```sql
SELECT
  -- Pass the params through explicitly
  {> some_select_query({fooId: id, badger: badger }) }
FROM
  some_table
```

some_select_query.sql

```sql
"some_table".*,
-- 'badger` was passed straight through
{ badger } AS "badger",
(
  SELECT
    row_to_json(*)
  FROM
    some_other_table
  WHERE
    -- note that 'id' is 'fooId' here
    some_other_table.id = {fooId}
  LIMIT
    1
) AS "subQueryData"
```

**Comment by orangemug** [comment]
@oliverbrooks I think this is already supported `{>path, optionalDataKey}` unless I'm missing something?

**Comment by orangemug** [comment]
Ok re-reading this I think I see the point you're trying to make. I'm thinking a better solution would be to add an additional `sql-stamp-hannibal` plugin that on require `{> ./some/partial.sql}` would also include a `{> ./some/partial.schema.json}` and run a strict validation using https://github.com/pearlshare/hannibal

**Comment by orangemug** [comment]
Maybe with an additional syntax in the require

 * `{>./path userId:user.id, ...data}`

Basically my main aim to to keep the SQL as clean as possible. The rules here are as follows

 * `key` - `key` is available in the partial
 * `foo:key` - `key` is available as `foo` in the partial
 * `...key` - `key` is assumed to be an object and expanded, For example if `key` was the object `{a: 1, b: 2}` the partial would be passed

All the values use `Object#assign` in the order they are passed. thoughts?

**Comment by orangemug** [comment]
Removed from [v0.7.0](https://github.com/orangemug/sql-stamp/milestones/v0.7.0) milestone, if we can agree on syntax I'll add back into `v0.8.0`

**Comment by orangemug** [comment]
With my current thinking the `README.md` would look something like 

 * `{>path, optionalDataKeys*}` - Require file from the templates, the data keys can take on the format
   * `{>path id:user.id}` - pass `user.id` as `id`
   * `{>path status}` - pass `status` as `status`
   * `{>path ...user}` - object spread, in the same way as javascript [ES6 spread operator](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Spread_operator)

**Comment by orangemug** [comment]
@oliverbrooks, @pasupulaphani can you add some use cases from your codebase?

**Comment by oliverbrooks** [comment]
The use cases are when sharing sql between queries. We often use short sql snippets to add presentation data to query results. For example a SQL template which loads details for an account.

The presentation sql snippet may look like this, where `userId` is the id of the current user:

```sql
"account"."id",
"account"."firstName",
(
  SELECT
    ROW_TO_JSON("friend".*)
  FROM
    "friend"
  WHERE
      ("friend"."toId" = "account"."id" AND "friend"."fromId" = {userId})
    OR
      ("friend"."fromId" = "account"."id" AND "friend"."toId" = {userId})
) AS "friendship"
```

the `userId` param might be used in the parent sql snippet for something else.

```sql
SELECT
  {> users/_present.sql} AS "friend"
FROM
  account
WHERE
  "account"."id" = {userId}
```
```js
sql.queries.userAndFriend({
  userId: id,
  currentUserId: req.user.id
});
```

So far we've got around this by being strict about use of variable names on a global scope. Would be nice to isolate the scope for each SQL snippet.

```sql
SELECT
  {> users/_present.sql, userId:currentUserId } AS "friend"
FROM
  account
WHERE
  "account"."id" = {userId}
```

The syntax `{> users/_present.sql, userId:currentUserId }` would map the `currentUserId` passed into the parent sql snippet as `userId` in the child snippet thus allowing snippets to be re-used safely without worrying about global naming of variables.

This is the same behaviour as rails partials:

```erb

<div>
  { render "my/partial", userId: accountId }
</div>
```

**Comment by orangemug** [comment]
Should we just use that syntax or also `...user` and `user`, see <https://github.com/orangemug/sql-stamp/issues/9#issuecomment-224187538>

**Comment by oliverbrooks** [comment]
@orangemug don't see anything wrong with doing that. 

Question I have is do you always need to explicitly pass everything into the partials or do passed params override globals?

**Comment by orangemug** [comment]
@oliverbrooks see <https://github.com/orangemug/sql-stamp/pull/11>. In this PR you always explicitly pass the params.

**Comment by oliverbrooks** [comment]
@orangemug I think it is right to do so. Globals? yuk.

Big breaking change but I think worth it.

**Comment by orangemug** [comment]
Published in [v0.8.0](https://github.com/orangemug/sql-stamp/tree/v0.8.0)
